_posts/2020-12-25-manual-gates-git-flow-in-a-wig.md

```diff
@@ -0,0 +1,145 @@
+---
+layout: post
+title: Manual Gates, or "GitFlow in a Wig"
+date: 2020-08-19
+tags: devops ci cd gitflow trunk
+author: Matt Tyler
+---
+
+Over last few years I've spent a lot of time working in the CI/CD salt mines. The simplest expression of a CI/CD pipeline is depicted below; we begin with pulling the source code, go through to a build and test stage, and deploy to an environment.
+
+<center><img src="/img/manual-gates/simple-1.png" /></center>
+<br/>
+Most teams will start with this. At some point, the team itself will decide there is too much risk deploying directly to production, and will decide they need an additional environment. Enter the test (or nonproduction, or development, or QC) stage. They put in a manual gate between deployment phases and everything is fine. The pipeline starts to get used in other projects and the world keeps turning.
+
+<center><img src="/img/manual-gates/simple-2.png" /></center>
+<br/>
+Then a new project comes along, and decides two stages isn't enough. They want another one. It could be for User Acceptance Testing. It could be for training. At any rate it gets shoved into the pipeline between the nonproduction and production stages and the world kicks on; although the teams have noticed that this project does not deploy to production as often.
+
+<center><img src="/img/manual-gates/simple-3.png" /></center>
+<br/>
+More time passes. Other teams decide they need another environment. Others add other types of verification steps into the pipeline. As they get more and more complicated, the likelihood of code getting through all the way without manual intervention decreases. Frustrated with increasing complexity, teams find ways to circument the pipeline in order to deploy manually. Incidents start to increase in frequency and duration. Engineers moving between teams are frustrated because the pipelines have wildly diverged.
+
+Sounding familiar?
```

Comment:
This introduction is good.  But I think it maybe leaves one thing unsaid: *what is the purpose of your pipeline*?

Teams are familiar with continuous delivery, and will say they practice it even if they don't really.
Teams think they know what CI/CD is/is for - and that's the underlying point of your post - that they may misunderstand the purpose.

So clearly stating your definition of the purpose of the pipeline is an important part of your argument.

Reply:
Good idea